CBinding/CLang/CLangManager.cs

```diff
@@ -0,0 +1,528 @@
+﻿using System;
+using CBinding;
+using ClangSharp;
+using MonoDevelop.Ide;
+using MonoDevelop.Projects;
+using System.Collections.Generic;
+using System.Threading;
+using MonoDevelop.Ide.Editor;
+using System.Runtime.InteropServices;
+using CBinding.Refactoring;
+using CBinding.Parser;
+using System.IO;
+using MonoDevelop.Ide.CodeCompletion;
+
+namespace CBinding
+{
+	/// <summary>
+	/// Class to manage clang Translation units, thread safe, but the dictionary exposed
+	/// and the translation units itself are not.
+	/// For more information see the field SyncRoot
+	/// </summary>
+	public class CLangManager : IDisposable
+	{
+		/// <summary>
+		/// The sync root.
+		/// I couldn't find any information about libclang's internal threading solutions,
+		/// the best i could find were some stackoverflow topic and a short mailing list discussion, where only assumptions were made.
+		/// Lock on this is needed because clangmanager handles getting cursors,
+		/// cursor references, and cursors between files sometimes are a mess and
+		/// can only identified by its USR, and e.g. finding a cursor by its USR
+		/// while reparsing is in progress in an other file could result in a fault.
+		/// </summary>
+		public readonly object SyncRoot = new object ();
+		CProject project;
+		CXIndex index;
+		Dictionary<string, CXTranslationUnit> translationUnits;
+
+		public Dictionary<string, CXTranslationUnit> TranslationUnits {
+			get {				
+				lock (SyncRoot) {
+					return translationUnits;
+				}
+			}
+		}
+
+		/// <summary>
+		/// Constructor
+		/// </summary>
+		/// <param name="proj">
+		/// A <see cref="CProject"/> reference: project which the manager manages
+		/// </param>
+		public CLangManager (CProject proj)
+		{
+			project = proj;
+			index = clang.createIndex (0, 0);
+			translationUnits = new Dictionary<string, CXTranslationUnit> ();
+			project.DefaultConfigurationChanged += CompilerArgumentsUpdate;
+		}
+
+		/// <summary>
+		/// Creates a new or gives back a previously created translation unit
+		/// </summary>
+		/// <param name="proj">
+		/// A <see cref="CProject"/> reference: the project which the TU is associated. Contains the compiler arguments, in its configurations
+		/// </param>
+		/// <param name="fileName">
+		/// A <see cref="string"/>: The filename associated with the translation unit. Basically the source file's name
+		/// </param>
+		/// <param name = "unsavedFiles">
+		/// A <see cref="CXUnsavedFile"/> array: array with the contents of unsaved files in IDE. Safe to be a null sized array - CDocumentParser.Parse reparses the TU with properly initialized unsaved files.
+		/// </param>
+		/// <returns>
+		/// A <see cref="CXTranslationUnit"/>: The Translation unit created
+		/// </returns>
+		public CXTranslationUnit createTranslationUnit(CProject proj, string fileName, CXUnsavedFile[] unsavedFiles) 
```

Comment:
And you can ignore the C# naming conventions for the DllImports - a 1:1 mapping from C to C# is fine.